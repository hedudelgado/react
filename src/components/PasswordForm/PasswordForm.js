import React from 'react';
import { object, func } from 'prop-types';
import Input from '../../layout/Input';
import Button from '../../layout/Button';
import { DEFAULT_VIEW } from '../App/appActions';

const PasswordForm = ({
  dictionary,
  errors,
  touched,
  handleSubmit,
  setEditView,
  handleChange,
  handleBlur,
}) => (
  <form
    name="password-form"
    onSubmit={handleSubmit}
  >
    <Input
      onBlur={handleBlur}
      onChange={handleChange}
      type="password"
      name="currentPassword"
      id="currentPassword"
      placeholder={dictionary['password.currentpassword']}
      error={dictionary[errors.currentPassword]}
      touched={touched.currentPassword}
    />
    <Input
      onBlur={handleBlur}
      onChange={handleChange}
      type="password"
      name="newPassword"
      id="newPassword"
      placeholder={dictionary['password.newpassword']}
      error={dictionary[errors.newPassword]}
      touched={touched.newPassword}
    />
    <Input
      onBlur={handleBlur}
      onChange={handleChange}
      type="password"
      name="confirmPassword"
      id="confirmPassword"
      placeholder={dictionary['password.confirmpassword']}
      error={dictionary[errors.confirmPassword]}
      touched={touched.confirmPassword}
    />
    <ul>
      <li>{dictionary['password.requirements.min']}</li>
      <li>{dictionary['password.requirements.number']}</li>
      <li>{dictionary['password.requirements.capital']}</li>
      <li>{dictionary['password.requirements.special']}</li>
    </ul>
    <Button
      className="wb-push--bottom wb-btn--full-width"
      id="save-changed"
      color="primary"
      type="submit"
    >
      {dictionary['password.button.update']}
    </Button>
    <Button
      data-test="closeForm"
      color="default"
      variant="text"
      className="underline block"
      fullWidth
      onClick={() => setEditView(DEFAULT_VIEW)}
    >
      {dictionary['password.button.cancel']}
    </Button>
  </form>
);

PasswordForm.propTypes = {
  dictionary: object.isRequired,
  errors: object.isRequired,
  touched: object.isRequired,
  handleSubmit: func.isRequired,
  setEditView: func.isRequired,
  handleChange: func.isRequired,
  handleBlur: func.isRequired,
};

export default PasswordForm;
