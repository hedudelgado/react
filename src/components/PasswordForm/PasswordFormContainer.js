import React, { PureComponent, createRef } from 'react';
import { func, object } from 'prop-types';
import { Formik } from 'formik';
import BaseStyle from '../../layout/BaseStyle';
import Typography from '../../layout/Typography';
import PasswordForm from './PasswordForm';
import Notification from '../../layout/Notification';
import { validateCurrentPassword, validateNewPassword, validateConfirmPassword } from '../../utils/validations';
import {
  DEFAULT_VIEW,
  PASSWORD_ERROR,
  PASSWORD_VIEW,
} from '../App/appActions';
import { apiErrorCodes } from '../../config.json';

const validatePasswordForm = (values) => {
  const errors = {};

  errors.currentPassword = validateCurrentPassword(values.currentPassword);
  errors.newPassword = validateNewPassword(values);
  errors.confirmPassword = validateConfirmPassword(values);

  Object.keys(errors).forEach((error) => {
    if (errors[error] === undefined) delete errors[error];
  });

  return errors;
};

export default class PasswordFormContainer extends PureComponent {
  constructor(props) {
    super(props);

    this.form = createRef();
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleSubmit(values, actions) {
    const {
      profile,
      setNotificationMessage,
      setEditView,
      updateProfile,
    } = this.props;
    const {
      newPassword,
      currentPassword: password,
    } = values;
    const profileRequest = { ...profile, password, newPassword };

    updateProfile(profileRequest).then(() => {
      setNotificationMessage(PASSWORD_VIEW);
      setEditView(DEFAULT_VIEW);
      actions.setSubmitting(false);
    }).catch(({ code }) => {
      const { INCORRECT_PASSWORD } = apiErrorCodes;
      if (code === INCORRECT_PASSWORD) {
        actions.setFieldError('currentPassword', 'password.error.incorrect');
      } else {
        setNotificationMessage(PASSWORD_ERROR);
      }
    });
  }

  render() {
    const {
      setEditView,
      dictionary,
      app: { notificationMessage },
    } = this.props;

    return (
      <BaseStyle component="section" className="component-container">
        {notificationMessage === PASSWORD_ERROR && (
          <Notification data-test="error-notification" priority={2} variant="error" ariaLive="polite" show>
            {dictionary['generic.error']}
          </Notification>
        )}
        <Typography component="h4" className="pi-color-greyscale-darkest">
          {dictionary['password.changepassword.title']}
        </Typography>

        <Formik
          data-test="password-form-wrapper"
          ref={this.form}
          initialValues={{ currentPassword: '', newPassword: '', confirmPassword: '' }}
          validate={validatePasswordForm}
          onSubmit={this.handleSubmit}
          render={formikProps => (
            <PasswordForm
              {...formikProps}
              data-test="password-form"
              dictionary={dictionary}
              setEditView={setEditView}
            />
          )}
        />
      </BaseStyle>
    );
  }
}

PasswordFormContainer.propTypes = {
  dictionary: object.isRequired,
  setEditView: func.isRequired,
  profile: object.isRequired,
  app: object.isRequired,
  setNotificationMessage: func.isRequired,
  updateProfile: func.isRequired,
};
