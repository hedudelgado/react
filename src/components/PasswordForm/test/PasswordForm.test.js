import React from 'react';
import { shallow } from 'enzyme';
import PasswordForm from '../PasswordForm';

describe('PasswordForm component', () => {
  let wrapper;
  const props = {
    dictionary: {
      'password.button.update': 'Update Password',
      'password.button.cancel': 'Cancel password update',
    },
    errors: {},
    touched: {},
    handleSubmit: jest.fn(),
    setEditView: jest.fn(),
    handleChange: jest.fn(),
    handleBlur: jest.fn(),
  };

  beforeEach(() => {
    wrapper = shallow(<PasswordForm {...props} />);
  });

  describe('renders the component elements', () => {
    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('follows the business logic', () => {
    beforeEach(() => {
      jest.clearAllMocks();
    });

    it('should go back to default view on Cancel', () => {
      wrapper.find('[data-test="closeForm"]').simulate('click');
      expect(props.setEditView).toHaveBeenCalled();
    });

    it('should submit the form', () => {
      wrapper.find('[name="password-form"]').simulate('submit');
      expect(props.handleSubmit).toHaveBeenCalledTimes(1);
    });

    it('should call the blur function on field Blur event', () => {
      wrapper.find('[name="currentPassword"]').simulate('blur');
      wrapper.find('[name="newPassword"]').simulate('blur');
      wrapper.find('[name="confirmPassword"]').simulate('blur');

      expect(props.handleBlur).toHaveBeenCalledTimes(3);
      expect(props.handleChange).toHaveBeenCalledTimes(0);
    });

    it('should call the change function on field Change event', () => {
      wrapper.find('[name="currentPassword"]').simulate('change');
      wrapper.find('[name="newPassword"]').simulate('change');
      wrapper.find('[name="confirmPassword"]').simulate('change');

      expect(props.handleBlur).toHaveBeenCalledTimes(0);
      expect(props.handleChange).toHaveBeenCalledTimes(3);
    });
  });
});
