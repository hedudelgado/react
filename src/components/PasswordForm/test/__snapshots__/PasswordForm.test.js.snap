// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordForm component renders the component elements should match snapshot 1`] = `
<form
  name="password-form"
  onSubmit={[MockFunction]}
>
  <Input
    id="currentPassword"
    mb="wb-push-middle--bottom"
    name="currentPassword"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    sessioncamhide={true}
    type="password"
  />
  <Input
    id="newPassword"
    mb="wb-push-middle--bottom"
    name="newPassword"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    sessioncamhide={true}
    type="password"
  />
  <Input
    id="confirmPassword"
    mb="wb-push-middle--bottom"
    name="confirmPassword"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    sessioncamhide={true}
    type="password"
  />
  <ul>
    <li />
    <li />
    <li />
    <li />
  </ul>
  <Button
    className="wb-push--bottom wb-btn--full-width"
    color="primary"
    fullWidth={false}
    href={null}
    id="save-changed"
    onClick={[Function]}
    submit={false}
    type="submit"
    variant="button"
  >
    Update Password
  </Button>
  <Button
    className="underline block"
    color="default"
    data-test="closeForm"
    fullWidth={true}
    href={null}
    onClick={[Function]}
    submit={false}
    variant="text"
  >
    Cancel password update
  </Button>
</form>
`;
